### ISSUE TITLE
Feat.: Support for "Advanced_Optimizers"

### ISSUE THREAD

**Issue by betterftr** [kohya-ss/musubi-tuner#598]
https://github.com/Koratahiu/Advanced_Optimizers

All support fused backward pass, Onetrainer already implemented them and they work really great!
(https://github.com/Nerogar/OneTrainer/pull/992)

**Comment by FurkanGozukara** [comment]
@betterftr do you know if any of them quality wise better? or all memory

I tested recently a lot in Kohya and Adafactor yielded best quality wise on FLUX

**Comment by betterftr** [comment]
I only tested them in Onetrainer (with Qwen edit) so quality could differ in a different codebase, and depends on many factors but I was satisfied with both the memory footprint and the quality

**Comment by kohya-ss** [comment]
Thank you for this! That seems to be promising. There doesn't seem to be any usage documentation in the repository, is there somewhere? Should I look at the OneTrainer code?

**Comment by betterftr** [comment]
I think yes, the most "challenging" part of implementing them is fused supp, since your flag for fused back pass only hooks for adafactor and these support it natively with:

```
def supports_fused_back_pass(self) -> bool:
        return True
```

and ofc --optimizer_args

So I guess your best bet is yea compare with the OT implementation
they have a wiki for them if it helps understanding the args a bit more:
https://github.com/Nerogar/OneTrainer/wiki/Advanced-Optimizers

**Comment by betterftr** [comment]
I did a ChatGpt Codex implementation, you can check it here, it works (note that this is modified version of your codebase, just mind the advanced optimizer implementation please :D) Basically I slapped the folder "Advanced_Optimizers" outside the full root directory of musubi-tuner (obviously not the best idea since there is a pip install for this)

https://drive.google.com/drive/folders/17XOUwaPDdReUN9TiFbkFMuZh7QQVDNCC?usp=sharing

<img width="1065" height="426" alt="Image" src="https://github.com/user-attachments/assets/9c142f92-ab0f-49a1-a080-d5c21a809ab0" />

<img width="485" height="358" alt="Image" src="https://github.com/user-attachments/assets/365eb1f8-8552-44ad-b3a9-4384204ff3a2" />
(wan ft bs 2)
command:
accelerate launch --num_cpu_threads_per_process 1 --mixed_precision bf16 wan_train.py --task t2v-A14B --dit C:/train/saves/runs/tanya_512-000110.safetensors --dataset_config C:/train/hunyuan_dataset.toml --sdpa --mixed_precision bf16 --learning_rate 5e-6 --gradient_checkpointing --max_data_loader_n_workers 2 --persistent_data_loader_workers --timestep_sampling shift --discrete_flow_shift 1.0 --max_train_epochs 150 --save_every_n_epochs 5 --sample_every_n_epochs 5 --seed 1234 --output_dir C:/train/saves --output_name tanya_512 --vae C:/train/ckpts/wan_2.1_vae.safetensors --t5 C:/train/ckpts/models_t5_umt5-xxl-enc-bf16.pth --sample_prompts "C:/train/sample/random_prompt.txt" --logging_dir "C:/train/logs" --lr_warmup_steps "50" --lr_scheduler "linear" --preserve_distribution_shape --min_timestep "0" --max_timestep "1000" --blocks_to_swap 29 --full_bf16 --mem_eff_save --optimizer_type Adopt_adv --max_grad_norm 0 --optimizer_args "nnmf_factor=True"

Tiny bit more vram usage than Adafactor with fused back pass, with that I could do batch size 8, with this 6-7, with res 512, I say thats a damn good compromise ^^

**Comment by Sarania** [comment]
Oh nice, with support for stochastic rounding for all of them too. Hype!

**Comment by sdbds** [comment]
Looks good, do we need to do any hook operations to support the feature?
